### PR TITLE
[feat] ✨ Multicall and adding the feature to cancel the campaign and recalculations of rewards

### PR DESCRIPTION
**Pull request by guillermoriv**
Adding the feature to recalculate the rewards based on the first stake and some test cases for this situation

### CODE REVIEW

**Review by mr-onadasky — CHANGES_REQUESTED** [review]
Very nice improvement!
There are a few questions I left.

Does this functionality also need to be added to the infinite staking pools?

File: contracts/RewardsPoolBase.sol [hunk 1]
```diff
@@ -35,6 +35,7 @@ contract RewardsPoolBase is Ownable {
 
     IERC20 public immutable stakingToken;
 
+    uint256 public originalTimestamp;
```
[the remark this answers is not in the document]
Comment (guillermoriv) [reply contracts/RewardsPoolBase.sol:38]: Yes, it will get use at least in the frontend to show the correct duration of the pool, and not the modified one for the calculations

[the remark this answers is not in the document]
Comment (guillermoriv) [reply contracts/RewardsPoolBase.sol:38]: And with regards to the infinitePools, I will take a look at this tomorrow morning and come back to you with an answer @Mr-ONADASKY

[the remark this answers is not in the document]
Comment (guillermoriv) [reply contracts/RewardsPoolBase.sol:38]: Yes indeed, this will be needed for the infinite pools as well, so I'll do the implementation there

File: contracts/infinite-pool/RewardsPoolBaseInfinite.sol [hunk 2]
```diff
@@ -168,21 +165,6 @@ contract RewardsPoolBaseInfinite is Ownable {
         emit Started(startTimestamp, endTimestamp, rewardPerSecond);
     }
 
-    /** @dev Cancels the scheduled start. Can only be done before the start.
```
Comment (Mr-ONADASKY) [inline contracts/infinite-pool/RewardsPoolBaseInfinite.sol:171]: Why is this removed?

File: test/RewardsPoolInfinite.ts [hunk 3]
```diff
@@ -287,27 +231,6 @@ describe('RewardsPoolBaseInfinite', () => {
       expect(await stakingToken.balanceOf(rewardsPoolBaseInfinite.address)).to.be.eq(amount);
     });
 
-    it('Should not be able to call onlyOwner functions', async () => {
```
Comment (Mr-ONADASKY) [inline test/RewardsPoolInfinite.ts:290]: Why is this test removed?

File: contracts/infinite-pool/RewardsPoolBaseInfinite.sol [hunk 4]
```diff
@@ -155,34 +156,17 @@ contract RewardsPoolBaseInfinite is Ownable {
...
         rewardPerSecond = _rewardPerSecond;
+        originalTimestamp = _startTimestamp;
         startTimestamp = _startTimestamp;
         endTimestamp = _endTimestamp;
         lastRewardTimestamp = _startTimestamp;
 
         emit Started(startTimestamp, endTimestamp, rewardPerSecond);
     }
 
-    /** @dev Cancels the scheduled start. Can only be done before the start.
```
Comment (Mr-ONADASKY) [inline contracts/infinite-pool/RewardsPoolBaseInfinite.sol:171]: Why is this removed?

File: test/RewardsPoolInfinite.ts (hunk 3 above)
Comment (Mr-ONADASKY) [inline test/RewardsPoolInfinite.ts:290]: Why is this removed?

**Review by mr-onadasky — COMMENTED** [review]
Looks good, but I'm not sure if some removals where  by accident or on purpose.
Aside from those it looks good

File: contracts/infinite-pool/RewardsPoolBaseInfinite.sol (hunk 2 above)
Comment (guillermoriv) [reply contracts/infinite-pool/RewardsPoolBaseInfinite.sol:171]: Following the discussion we had yesterday in the meeting, the cancel shouldn't be a thing in infinitePools, so I'll close this one @Mr-ONADASKY

File: test/RewardsPoolInfinite.ts (hunk 3 above)
Comment (guillermoriv) [reply test/RewardsPoolInfinite.ts:290]: Same thing as mention above, so i'll close this one

**Review by mr-onadasky — APPROVED** [review]
LGTM

File: contracts/RewardsPoolBase.sol [hunk 5]
```diff
@@ -323,20 +344,27 @@ contract RewardsPoolBase is Ownable {
...
-
         uint256 secondsSinceLastReward = applicableTimestamp - lastRewardTimestamp;
+        uint256 rewardsTokensLength = rewardsTokens.length;
 
         if (secondsSinceLastReward == 0) {
             return;
         }
 
         if (totalStaked == 0) {
+            uint256[] memory _rewardPerSecond = new uint256[](rewardsTokensLength);
```
Comment (boukestam-allianceblock) [inline contracts/RewardsPoolBase.sol:355]: This breaks in the following scenario:

You have a 30 day campaign, somebody stakes at 0 days, then withdraws at 5 days. So now there is 0 totalStaked again. Then at 10 days he stakes again, so now it will take all rewards and spread it over the last 20 days. But actually some rewards were already used in the first 5 days. No there is not enough rewards, and it will break.

We need to make sure it only can do this the first time that totalStaked is 0

**Review by boukestam-allianceblock — CHANGES_REQUESTED** [review]

**Review by mr-onadasky — COMMENTED** [review]
Changes look okay.
I've added a few questions
